beacon_node/beacon_chain/src/block_verification.rs

```diff
@@ -855,17 +895,16 @@ impl<T: BeaconChainTypes> GossipVerifiedBlock<T> {
         //
         // It's important to double-check that the proposer still hasn't been observed so we don't
         // have a race-condition when verifying two blocks simultaneously.
-        if chain
+        match chain
             .observed_block_producers
             .write()
-            .observe_proposer(block.message())
+            .observe_proposal(block_root, block.message())
             .map_err(|e| BlockError::BeaconChainError(e.into()))?
         {
-            return Err(BlockError::RepeatProposal {
-                proposer: block.message().proposer_index(),
-                slot: block.slot(),
-            });
-        }
+            SeenBlock::Slashable => return Err(BlockError::SlashableProposal),
+            SeenBlock::Duplicate => return Err(BlockError::BlockIsAlreadyKnown),
+            _ => {}
```

Comment:
may as well round out the match here, so variants added in future must be handled

```suggestion
            SeenBlock::UniqueNonSlashable => {}
```